mm2src/lp_ordermatch/simple_market_maker.rs

```diff
@@ -0,0 +1,734 @@
+    rates: &RateInfos,
+    cfg: &SimpleCoinMarketMakerCfg,
+    key_trade_pair: String,
+) -> OrderProcessingResult {
+    if rates.base_provider == Provider::Unknown || rates.rel_provider == Provider::Unknown {
+        warn!("rates from provider are Unknown - skipping for {}", key_trade_pair);
+        return MmError::err(OrderProcessingError::ProviderUnknown);
+    }
+
+    if rates.price.is_zero() {
+        warn!("price from provider is zero - skipping for {}", key_trade_pair);
+        return MmError::err(OrderProcessingError::PriceIsZero);
+    }
+
+    if rates.last_updated_timestamp == 0 {
+        warn!(
+            "last updated price timestamp is invalid - skipping for {}",
+            key_trade_pair
+        );
+        return MmError::err(OrderProcessingError::LastUpdatedTimestampInvalid);
+    }
+
+    // Elapsed validity is the field defined in the cfg or 5 min by default (300 sec)
+    let time_diff = rates.retrieve_elapsed_times();
+    let elapsed = match time_diff.elapsed() {
+        Ok(elapsed) => elapsed.as_secs_f64(),
+        Err(_) => return MmError::err(OrderProcessingError::PriceElapsedValidityUntreatable),
+    };
+    let elapsed_validity = cfg.price_elapsed_validity.unwrap_or(300.0);
+
+    if elapsed > elapsed_validity {
+        warn!(
+            "last updated price timestamp elapsed {} is more than the elapsed validity {} - skipping for {}",
+            elapsed, elapsed_validity, key_trade_pair,
+        );
+        return MmError::err(OrderProcessingError::PriceElapsedValidityExpired);
+    }
+    info!("elapsed since last price update: {} secs", elapsed);
+    Ok(true)
+}
+
+async fn prepare_order(
+    rates: RateInfos,
+    cfg: SimpleCoinMarketMakerCfg,
+    key_trade_pair: String,
+    ctx: &MmArc,
+) -> OrderPreparationResult {
+    checks_order_prerequisites(&rates, &cfg, key_trade_pair.clone()).await?;
+    let base_coin = lp_coinfind(ctx, cfg.base.as_str())
+        .await?
+        .ok_or_else(|| MmError::new(OrderProcessingError::AssetNotEnabled))?;
+    let base_balance = base_coin.get_non_zero_balance().compat().await?;
+    lp_coinfind(ctx, cfg.rel.as_str())
+        .await?
+        .ok_or_else(|| MmError::new(OrderProcessingError::AssetNotEnabled))?;
+
+    info!("balance for {} is {}", cfg.base, base_balance);
+
+    let mut calculated_price = rates.price * cfg.spread.clone();
+    info!("calculated price is: {}", calculated_price);
+    if cfg.check_last_bidirectional_trade_thresh_hold.unwrap_or(false) {
+        calculated_price = vwap_calculator(calculated_price.clone(), ctx, &cfg).await?;
+    }
+
+    let volume = match cfg.balance_percent {
+        Some(balance_percent) => balance_percent * base_balance.clone(),
+        None => MmNumber::default(),
+    };
+
+    let min_vol: Option<MmNumber> = match cfg.min_volume {
+        Some(min_volume) => {
+            if cfg.max.unwrap_or(false) {
+                Some(min_volume * base_balance.clone())
```

Comment:
The naming is still confusing here.

Reply:
> The naming is still confusing here.

I renamed `cfg.min_volume -> cfg.min_volume_percentage` and `Some(min_volume)` -> `Some(min_volume_percentage)`